nimsuggest/tests/tsug_regression.nim

```diff
@@ -17,13 +17,17 @@ proc main =
   map0.#[!]#
 
 discard """
-disabled:true
 $nimsuggest --tester $file
 >sug $1
-sug;;skProc;;tables.getOrDefault;;proc (t: Table[getOrDefault.A, getOrDefault.B], key: A): B;;$lib/pure/collections/tables.nim;;178;;5;;"";;100;;None
-sug;;skProc;;tables.hasKey;;proc (t: Table[hasKey.A, hasKey.B], key: A): bool;;$lib/pure/collections/tables.nim;;233;;5;;"returns true iff `key` is in the table `t`.";;100;;None
-sug;;skProc;;tables.add;;proc (t: var Table[add.A, add.B], key: A, val: B);;$lib/pure/collections/tables.nim;;309;;5;;"puts a new (key, value)-pair into `t` even if ``t[key]`` already exists.";;100;;None
-sug;;skIterator;;tables.allValues;;iterator (t: Table[allValues.A, allValues.B], key: A): B{.inline.};;$lib/pure/collections/tables.nim;;225;;9;;"iterates over any value in the table `t` that belongs to the given `key`.";;100;;None
-sug;;skProc;;tables.clear;;proc (t: var Table[clear.A, clear.B]);;$lib/pure/collections/tables.nim;;121;;5;;"Resets the table so that it is empty.";;100;;None
+sug;;skProc;;tables.hasKey;;proc (t: Table[hasKey.A, hasKey.B], key: A): bool;;*/lib/pure/collections/tables.nim;;374;;5;;"Returns true if*";;100;;None
+sug;;skProc;;tables.add;;proc (t: var Table[add.A, add.B], key: A, val: sink B);;*/lib/pure/collections/tables.nim;;505;;5;;"Puts a new*";;100;;None
+sug;;skIterator;;tables.allValues;;iterator (t: Table[allValues.A, allValues.B], key: A): B{.inline.};;*/lib/pure/collections/tables.nim;;769;;9;;"Iterates over any*";;100;;None
+sug;;skProc;;tables.clear;;proc (t: var Table[clear.A, clear.B]);;*/lib/pure/collections/tables.nim;;567;;5;;"Resets the table so that it is empty.*";;100;;None
+sug;;skProc;;tables.contains;;proc (t: Table[contains.A, contains.B], key: A): bool;;*/lib/pure/collections/tables.nim;;392;;5;;"Alias of `hasKey*";;100;;None
 *
 """
+
+# Open Questions:
+# * Why did hasKey sort above all?
+# * Why did getOrDefault drop off?
+# * Why does a deprecated procedures like `add` rank high?
```

Comment:
see implementation:
```nim
proc suggestEverything(c: PContext, n, f: PNode, outputs: var Suggestions) =
  # do not produce too many symbols:
  for (it, scopeN, isLocal) in allSyms(c):
    var pm: PrefixMatch
    if filterSym(it, f, pm):
      outputs.add(symToSuggest(c.config, it, isLocal = isLocal, ideSug, n.info, 0, pm,
                                c.inTypeContext > 0, scopeN))
```
PR welcome to improve it using heuristics you suggest:
first, generate list of all acceptable candidates and add them to a priority queue, using same approach as what I'm doing in https://github.com/nim-lang/Nim/pull/16067; the priority in the queue can depend on scope (as done in that PR) plus heuristics (eg deprecated ranks lower, as you suggest)

then do walk the priority queue to get results in sorted order. Alternatively, using seq + sort instead of a priority queue is ok too, that's not the main point

Reply:
Turns out `suggestEverything` is called much of the time, but not all the time. In this particular case the code ends up in `sugExpr` and since we have a `nkDotExpr`, it takes the `suggestFieldAccess` path and ends up in `suggestObject` and `suggestOperations`.

Any how, `getOrDefault` do in fact show up, just further down the list. While I was poking about, I found `cmpSuggestions` which compares `Suggest` based on various factors, one of which is `quality` (`Range[0..100]`). I think I should either:

* have `Suggest` note whether something is deprecated and factor it into the comparator directly
* have quality by way of the `getQuality` proc lower the quality rating

I'm not a fan of adding more state, but the former is most appealing as it means quality remains a sort of measure of relevance/closeness, while deprecation feels more like a tie breaker akin to boosting. NB. I'm borrowing terminology from fulltext indexing/search engines such as Elasticsearch when I say relevance and boosting.

Whatever the change, I'm inclined to make that in a follow-up PR, want to keep this focused on tests.

Reply:
I don't care much how you implement it, but I completely agree that deprecated symbols should either not be suggested at all or be of low priority.